### ISSUE TITLE
Segmentation fault when attaching callback to another lws thread

### ISSUE THREAD

**Issue by hqfrank** [warmcat/libwebsockets#2215]
I am working on a project that needs secure stream to be running in the other thread rather than the main thread.
My design idea is to create and run the lws event loop in a new thread, then attach callbacks from the main thread to the lws thread.

I refer to the minimum example minimal-http-client-attach.

In my source code (using C++)

```
std::mutex ssmutex;
struct lws_context  *context;
/* definition of myss_t, mysstx, myssrx, myssstate */
...
static int
  lws_attach_with_thread_locking(struct lws_context *context, int tsi, lws_attach_cb_t cb, lws_system_states_t state, void *opaque, struct lws_attach_item **get) {
    int n;
    printf("%s\n", __func__);
    ssmutex.lock();
    n = __lws_system_attach(context, tsi, cb, state, opaque, get);
    ssmutex.unlock();
    return n;
  }

  lws_system_ops_t ops = {
    nullptr,
    nullptr,
    lws_attach_with_thread_locking,
  };
  
  void
  create_stream_attach_callback(struct lws_context *context, int tsi, void *opaque) {
    printf("%s\n", __func__);
    lws_ss_info_t                     ssi;
    memset(&ssi, 0, sizeof ssi);
    ssi.handle_offset           = offsetof(myss_t, ss);
    ssi.opaque_user_data_offset = offsetof(myss_t, opaque_data);
    ssi.rx                      = myss_rx;
    ssi.tx                      = myss_tx;
    ssi.state                   = myss_state;
    ssi.user_alloc              = sizeof(myss_t);
    ssi.streamtype              = "spam";
    printf("Creating ss type: %s\n", ssi.streamtype);
    if (lws_sspc_create(context, 0, &ssi, NULL, &myss_handle, NULL, NULL)) {
      printf("%s: failed to create secure stream\n", __func__);
    }
    return;
  }

  void myfunc() {
    printf("%s\n", __func__);
    int n = 0;
    while (!context && n++ < 30)
      std::this_thread::sleep_for(std::chrono::milliseconds(10));;

    if (!context) {
      lwsl_err("%s: context didn't start\n", __func__);
    }

    printf("Starts to attach the create_stream_attach_callback\n");
    /*
     * From our different, non event loop thread, ask for our attach
     * callback to get called when lws system state is OPERATIONAL
     */
    lws_system_get_ops(context)->attach(context, 0, create_stream_attach_callback, LWS_SYSTATE_OPERATIONAL, NULL, NULL);
}
```

After I run the code, it stops right at "lws_system_get_ops(context)->attach(context, 0, create_stream_attach_callback, LWS_SYSTATE_OPERATIONAL, NULL, NULL);".

gdb told me that "Thread 1 "myprogram" received signal SIGSEGV, Segmentation fault." at specifically the above "attach" call.

I do not see any printout of "create_stream_attach_callback" or "lws_attach_with_thread_locking", so I suspect the illegal address is related to the variables/arguments we see in that specific line.

Any suggestion on debugging?

**Comment by lws-team** [comment]
The situation `__lws_system_attach()` was designed for is a static firmware that has some other manager "process" that initially spawns generic application threads... and some of those threads then wish to drop their original dedicated thread identity and instead become part of the lws event loop and run completely from that.

They can call the lws_system attach machinery to schedule callbacks from lws thread context, and afterwards, exit their original spawning thread, freeing up the related stack and thread internal allocations from their original thread.  Afterwards, the code in the application will be executed wholly from lws event loop context, the original thread the application started with is destroyed.

It's basically a trick to one-time take over ownership of thread contexts for chunks of application code cleanly, collecting them in the lws event loop thread instead.  The advantage, especially as the number of applications goes up, is saving resources compared to individual threads, and allow direct participation in the central lws event loop.

If that matches your situation, then check that you pass in your system ops struct at context creation time (the case that `lws_system_get_ops(context)` is NULL)

https://libwebsockets.org/git/libwebsockets/tree/minimal-examples/http-client/minimal-http-client-attach/minimal-http-client-attach.c?h=main#n189

You can also log, eg, `lwsl_notice("%s: ops %p\n", __func__, lws_system_get_ops(context));` before the code to check things you are suspicious about.

Also I don't know what you're doing about encapsulating this in any class, but if so you have to mark the callbacks to be called by C as `static` members, indicating they won't pass any `this`.

If actually that description of the attach stuff doesn't match your situation, like you don't want to lose the separate thread, the lws attach stuff isn't the answer and you'll need a different approach.

**Comment by lws-team** [comment]
>  if (lws_sspc_create(context, 0, &ssi, NULL, &myss_handle, NULL, NULL)) {

If you're using SSPC to the SS proxy, then it's unlikely the attach stuff is what you want, it's designed for a static firmware like for RTOS.

If the idea is your client has a "comms thread", it should be OK to just instantiate the lws context in the thread you want to use it from, it doesn't have any dependency on the lws_context and the event loop being started in your "main" initial thread.  Its restriction is that it's not threadsafe itself, the lws_ss apis can only be called from the event loop thread context, whichever thread that is.

**Comment by hqfrank** [comment]
Thanks for the advice on debugging.

As for my use case, I don't think it fits the description you suggested. 

What I would like to do is an experimental work that using SS as the transport layer of a blocking RPC framework. Therefore, in my application, the calls are blocking, such as "write" or "read"; however, SS and lws are not blocking. When the event loop starts, the while loop won't exit, which means if it is within my program's main thread, the main thread cannot do anything outside the event loop. That is the reason I want the lws event loop to run outside the main thread.

I plan to use the shared variables between the main thread and the lws thread so that the data between the two threads can be buffered sequentially. I also plan to write code in the main program to bridge the blocking calls and the non-blocking event loop *tx *rx callbacks.

Is it possible to use SS/lws in this way?

I considered the absorption of the main program by the lws thread, however, in my use case, the main program should be able to generate a new request such as tx() arbitrarily, which means the requests cannot be pre-scheduled by an event loop style logic. Therefore, the main thread should be able to schedule a new event into the event loop at any time (the scheduled event does not need to be executed immediately).

Do I need to start a new issue? Or we can keep the discussion in this issue?

**Comment by lws-team** [comment]
All event loops are by their nature nonblocking, it's not an lws or SS thing.  Everything based around libuv, libev, libevent, glib etc all take the same nonblocking, serialized, Async approach.  It's popular because with blocking, every IO wait costs you a new thread, which is often only existing to provide something for the blocking semantics to block until your write is done or whatever.

> When the event loop starts, the while loop won't exit, which means if it is within my program's main thread, the main thread cannot do anything outside the event loop.

Those two things don't quite follow, you can schedule any number of arbitrary things to happen from inside the event loop, at us resolution, and if what you want stuff to happen in response to events that occur on an fd or socket, you can also bind those inside the same event wait.  You can schedule an arbitrary callback at +1us before starting the loop, the callback will run shortly after the event loop starts and you can further schedule whatever you wanted to do in there.

> I plan to use the shared variables between the main thread and the lws thread so that the data between the two threads can be buffered sequentially. I also plan to write code in the main program to bridge the blocking calls and the non-blocking event loop *tx *rx callbacks.

You can put blocking semantics on top of SS API if you want to, but it should do it by letting the event loop stuff do its thing, and use synchronization objects in the shared area to control making the calling thread block.  The existing scheme is like it is because it only invites the user to code produce output when it can be sent on immediately, ie, the sink controls output production naturally.  This is how network programming is, the source may feel that it's the critical piece of the puzzle with data to send, but in reality, the state of the link, intermediaries and sink are in charge of what data may be sent.  If you have a scheme where you offer random writes, but only buffer one chunk before blocking the source and unblocking after it's completely sent on its way, it should be OK.

> I considered the absorption of the main program by the lws thread, however, in my use case, the main program should be able to generate a new request such as tx() arbitrarily, which means the requests cannot be pre-scheduled by an event loop style logic. 

While the event loops all understand timers and so you can schedule things, their main job is to manage sets of fds and produce events on writeability (POLLOUT) or readability (POLLIN).  So data arriving and being able to send on sockets / fds are the native events understood by the event loop.

Lws has one threadsafe api, lws_cancel_service(), other threads may call this to force an event (usually on a pipe, or pair of UDP sockets) that creates an event on the event loop to wake it, this allows other threads to wake the event loop from its wait and usually get it to take a look at a memory area shared between the threads to find out what you want it to do (eg, schedule a tx from event loop side).

**Comment by hqfrank** [comment]
So I guess you agree that there should be two threads in my program: one is the main program thread, and the other is the lws thread.

After the lws event loop is running in the lws thread, I would like to create a secure stream with the endpoint "XXX" set. "XXX" is unknown before the lws event loop starts, and it is for example a url received by the main thread from another program in the system. When "XXX" is available, how should the main thread create the secure stream in the lws thread then? I assume I can use the *attach() api to attach a callback to the lws context, and within the callback, "XXX" is read and set as the endpoint metadata of a SS type, and the SS is created accordingly.

When the secure stream is created, after the connection is established, the main program (both main thread and lws thread) does not have any user data to tx to the endpoint. The main thread will wait and receive the to be txed user data from other programs in the system. When data is prepared, the main thread needs to request a tx in SS, so that data can be transmitted to the endpoint. How to do that? Any API allows the main thread to schedule a "request_tx" into the lws event loop? When the event happens, the *tx() callback defined will be called and the prepared user data will be read and sent to the endpoint via SS.

**Comment by lws-team** [comment]
> So I guess you agree that there should be two threads in my program: one is the main program thread, and the other is the lws thread.

I don't really know what you're trying to do enough to judge it... I'm talking about multiple threads because you say you want blocking semantics, so there has to be at least two threads then, one to do the work while the other is blocked.  It's generally better to not do that, since eg, it doesn't scale downwards for constrained devices, but for whatever reasons, that might be what you have to do.

There's no question you definitely "can" do it with the synchronization object scheme I described above, plus one level of tx buffering.

> *attach() api 

No that's for a really specific usecase involving RTOS, you don't want that AFAICS.  It's job is to bind the thread it's called on to the event loop so you can destroy the original thread.  But you want multiple threads each with their own life going on.

Instead of that if the question is "how do I... from a foreign (non-lws event loop) thread...?" the default answer is describe what you want in a shared region (with mutex locking) and call `lws_cancel_service()` from the foreign thread, eg

https://libwebsockets.org/git/libwebsockets/tree/minimal-examples/ws-server/minimal-ws-server-threads/protocol_lws_minimal.c?h=main#n124

The lws event wait will immediately wake and broadcast a callback event to lws protocols `LWS_CALLBACK_EVENT_WAIT_CANCELLED`, like this

https://libwebsockets.org/git/libwebsockets/tree/minimal-examples/ws-server/minimal-ws-server-threads/protocol_lws_minimal.c?h=main#n266

... from there, you are in the lws event loop thread context, you can lock the shared region and have a look at what's wanted to happen.  The other thread can prepare whole arguments for tasks in there like SS creation, and then get the event loop thread to carry it out.

This is very reliable, it's OK with many foreign threads going, works on all the variations of event libs etc.

> The main thread will wait and receive the to be txed user data from other programs in the system

Well... that sounds a lot like a proxy.  It's a lot easier if you can bring both sides of the proxy into the same event loop.  If you can't, well, OK.

> Any API allows the main thread to schedule a "request_tx" into the lws event loop?

The general approach is use a shared area and `lws_cancel_service()` to nudge the lws event loop to fulfil whatever you wanted from that side.
